File: Arrays/MissingNumber.js
// Find the missing number from the range 0 to n

// Time Complexity  O(n)
// Space Complexity O(1)

var missingNumber = function(nums) {
    const expectedSum = (nums.length * (nums.length + 1)) / 2;
    const sum = nums.reduce((accum, num) => accum + num, 0);
    return expectedSum - sum;
};

// bit manipulation
var missingNumber = function(nums) {
    let ans;
    for(let i = 0; i < nums.length; i++){
        ans = ans ^ nums[i];
    }
    for(let i = 0; i <= nums.length; i++){
        ans = ans ^ i;
    }
    return ans
};

File: Arrays/MissingNumber.java
// Find the missing number from the range 0 to n

// Time Complexity  O(n)
// Space Complexity O(1)

public int missingNumber(int[] nums) {
    int expectedSum = (nums.length * (nums.length + 1)) / 2;
    int sum = 0;
    for(int i = 0; i < nums.length; i++){
        sum += nums[i];
    }
    return expectedSum - sum;
}

// bit manipulation

public int missingNumber(int[] nums) {
    int ans = 0;
    for(int i = 0; i < nums.length; i++){
        ans ^= nums[i];
    }
    for(int i = 0; i <= nums.length; i++){
        ans ^= i;
    }
    return ans;
}
